File: nim/difference-of-squares/difference_of_squares.nim
import future, math, sequtils

proc sum(ns: seq[int]): int =
  ns.foldl(a + b)

proc squared(n: int): int =
  n.float.pow(2.0).int

proc square_of_sum*(n: int): int =
  lc[ x | ( x <- 0..n ), int ].sum.squared

proc sum_of_squares*(n: int): int =
  lc[ x.squared | ( x <- 0..n ), int ].sum

proc difference*(n: int): int =
  square_of_sum(n) - sum_of_squares(n)

File: ocaml/difference-of-squares/difference_of_squares.ml
let identity n = n
let squared n = n * n

let calculate n each_fn total_fn =
  let rec map_sum = function
      0 -> 0
    | x -> (each_fn x) + (map_sum (x - 1))
  in
  total_fn (map_sum n)

let square_of_sum n = calculate n identity squared
let sum_of_squares n = calculate n squared identity
let difference_of_squares n = (square_of_sum n) - (sum_of_squares n)
